### PR TITLE
gadget, bootloader: preserve managed boot assets during gadget updates

### PR DESCRIPTION
**Pull request by bboozzoo**
When updating gadget assets, make sure to preserve managed boot assets if those are in place.

### CODE REVIEW

File: gadget/mountedfilesystem.go [hunk 1]
```diff
@@ -299,6 +307,41 @@ func fsStructBackupPath(backupDir string, ps *LaidOutStructure) string {
 	return filepath.Join(backupDir, fmt.Sprintf("struct-%v", ps.Index))
 }
 
+func maybePreserveManagedBootAssets(mountPoint string, ps *LaidOutStructure) ([]string, error) {
+	if ps.Role != SystemSeed && ps.Role != SystemBoot {
+		return nil, nil
+	}
+
+	opts := &bootloader.Options{
```
Comment (zyga) [inline gadget/mountedfilesystem.go:315]: I think I understand this but I think it warrants a comment. (I'm talking about the Recovery and NoSlashBoot options set below).

File: gadget/mountedfilesystem_test.go [hunk 2]
```diff
@@ -2408,3 +2409,136 @@ func (s *mountedfilesystemTestSuite) TestMountedUpdaterNonFilePreserveError(c *C
...
+}
+
+func (s *mountedfilesystemTestSuite) TestMountedUpdaterGrubSeedAssetsManaged(c *C) {
+	managed := true
+	preserved := true
+	s.testMountedUpdaterGrubBootAssets(c, managed, gadget.SystemSeed, preserved)
+}
+
+func (s *mountedfilesystemTestSuite) TestMountedUpdaterGrubBootAssetsNotManaged(c *C) {
+	notManaged := false
```
Comment (zyga) [inline gadget/mountedfilesystem_test.go:2510]: This double negative is somewhat confusing. I would go with just managed := false and preserved := false as it retains the flow and sense of the tests above it.

Comment (bboozzoo) [reply gadget/mountedfilesystem_test.go:2510]: Renamed

File: gadget/mountedfilesystem.go (hunk 1 above)
Comment (bboozzoo) [reply gadget/mountedfilesystem.go:315]: Added a comment

File: gadget/mountedfilesystem.go [hunk 3]
```diff
@@ -299,6 +307,43 @@ func fsStructBackupPath(backupDir string, ps *LaidOutStructure) string {
...
+		return nil, err
+	}
+
+	mbl, ok := bl.(bootloader.ManagedAssetsBootloader)
+	if !ok {
+		// bootloader implementation does not support managing its
+		// assets
+		return nil, nil
+	}
+	managed, err := mbl.IsCurrentlyManaged()
```
Comment (pedronis) [inline gadget/mountedfilesystem.go:336]: when we install uc20 the first time what happens here, we get false and then we overwrite these later?

File: bootloader/bootloader.go [hunk 4]
```diff
@@ -144,6 +144,9 @@ type ExtractedRunKernelImageBootloader interface {
 type ManagedAssetsBootloader interface {
 	// IsCurrentlyManaged returns true when the on disk boot assets are managed.
 	IsCurrentlyManaged() (bool, error)
+	// BootAssets returns a list of boot assets managed by the bootloader in
+	// the boot filesystem.
+	BootAssets() []string
```
Comment (pedronis) [inline bootloader/bootloader.go:149]: this should be called ManagedAssets I think, to match the other method and to make clear that it lists only the managed ones, not all the assets related to booting with this bootloader

File: gadget/mountedfilesystem.go [hunk 5]
```diff
@@ -265,6 +266,7 @@ type mountedFilesystemUpdater struct {
 	*MountedFilesystemWriter
 	backupDir  string
 	mountPoint string
+	bootAssets []string
```
Comment (pedronis) [inline gadget/mountedfilesystem.go:269]: managedBootAsserts perhaps?

File: gadget/mountedfilesystem.go (hunk 3 above)
Comment (bboozzoo) [reply gadget/mountedfilesystem.go:336]: In install mode we'll write out the boot config. Then on first boot in run mode, the update-gadget-assets task is a nop until seeding is complete.

Comment (bboozzoo) [reply gadget/mountedfilesystem.go:336]: As discussed on IRC, I'll add a todo in the writer code and a note in #8930 that the InstallBootConfig must come last.

**Pull request merged**